File: Other/Review/Phase 2/Review2/kmp/kmp.swift
func KMPSearch(text: String, pattern: String) -> Int {
    if pattern.isEmpty {
        return 0
    }
    
    let textArray = Array(text)
    let patternArray = Array(pattern)
    
    func findLSP(pattern: [Character]) -> [Int] {
        var lsp = [Int](repeating: 0, count: pattern.count)
        var j = 0
        lsp[0] = 0
        
        for i in 1..<pattern.count {
            while j > 0 && pattern[i] != pattern[j] {
                j = lsp[j - 1]
            }
            if pattern[i] == pattern[j] {
                j += 1
                
            } 
            lsp[i] = j
        }
        
        return lsp
    }
    
    let lsp = findLSP(pattern: patternArray)
    var j = 0
    
    for i in 0..<textArray.count {
        while j > 0 && textArray[i] != patternArray[j] {
            j = lsp[j - 1]
        }
        if textArray[i] == patternArray[j] {
            if j == patternArray.count - 1 {
                return i - j
            }
            j += 1
        }
    }
    
    return -1
}

// test
let res = KMPSearch(text: "AABABABABC", pattern: "ABABC")
print(res) // 5

File: Other/Review/Phase 8/Review2/Part2/kmp/kmp.go
package main

import (
	"fmt"
)

func KMPSearch(text string, pattern string) int {
	if len(pattern) == 0 {
		return 0
	}

	var findLSP func(pattern string) []int
	findLSP = func(pattern string) []int {
		lsp := make([]int, len(pattern))
		lsp[0] = 0
		j := 0

		for i := 1; i < len(pattern); i++ {
			for j > 0 && pattern[i] != pattern[j] {
				j = lsp[j-1]
			}
			if pattern[i] == pattern[j] {
				j++
			}
			lsp[i] = j
		}

		return lsp
	}

	lsp := findLSP(pattern)
	j := 0

	for i := 0; i < len(text); i++ {
		for j > 0 && text[i] != pattern[j] {
			j = lsp[j-1]
		}
		if text[i] == pattern[j] {
			if j == len(pattern)-1 {
				return i - j
			}
			j++
		}
	}

	return -1
}

// test
func main() {
	res := KMPSearch("AABABABABC", "ABABC")
	fmt.Println(res) // 5
}
